model/inference.py

```diff
@@ -0,0 +1,171 @@
+
+    def create_user_dataframe(self, user):
+        return pd.DataFrame([user], columns=self.user_columns)
+
+    def preprocessing_user(self, user_df, user_mean, user_std):
+        if "updated" in user_df.columns:
+            age = (
+                pd.to_datetime(user_df.loc[:, "updated"])
+                - pd.to_datetime(user_df.loc[:, "created_at"]).dt.tz_localize(None)
+            ) / np.timedelta64(1, "Y")
+        else:
+            age = (
+                pd.to_datetime(pd.to_datetime("today"))
+                - pd.to_datetime(user_df.loc[:, "created_at"]).dt.tz_localize(None)
+            ) / np.timedelta64(1, "Y")
+        user_df["tweet_freq"] = user_df["statuses_count"] / age
+        user_df["followers_growth_rate"] = user_df["followers_count"] / age
+        user_df["friends_growth_rate"] = user_df["friends_count"] / age
+        user_df["favourites_growth_rate"] = user_df["favourites_count"] / age
+        user_df["listed_growth_rate"] = user_df["listed_count"] / age
+        user_df["followers_friends_ratio"] = user_df["followers_count"] / np.maximum(
+            user_df["friends_count"], 1
+        )
+        user_df["screen_name_length"] = user_df["screen_name"].str.len()
+        user_df["num_digits_in_screen_name"] = user_df["screen_name"].str.count(r"\d")
+        user_df["name_length"] = user_df["name"].str.len()
+        user_df["num_digits_in_name"] = user_df["name"].str.count(r"\d")
+        user_df["description_length"] = user_df["description"].str.len()
+        user_df.replace({False: 0, True: 1}, inplace=True)
+        user_df = user_df.select_dtypes("number").fillna(0.0)
+        return (user_df - user_mean) / user_std
+
+    def preprocessing_tweet(self, row):
+        rowlist = str(row).split()
+        rowlist = [word.strip() for word in rowlist]
+        rowlist = [
+            word if not word.strip().startswith("#") else "hashtagtag"
+            for word in rowlist
+        ]
+        rowlist = [
+            word if not word.strip().startswith("@") else "usertag" for word in rowlist
+        ]
+        rowlist = [word.lower() for word in rowlist]
+        rowlist = [re.sub(_URL_PATTERN, "urltag", word) for word in rowlist]
+        return " ".join(rowlist)
+
+    def vectorizing_tweet(self, tweets):
+        v = self.vectorizer.transform(tweets)
+        return v.A[np.newaxis, :]
+
+    def load_model(self, path, model_dict):
+        model = SOBOG(gpu=0, **model_dict)
+        model.load_state_dict(torch.load(path, map_location=torch.device("cpu")))
+        model.eval()
+        return model
+
+    def load_vectorizer(self, path):
+        with open(path, "rb") as fin:
+            vect = pickle.load(fin)
+        return vect
+
+    def generate_adj_matrix(self, tweet_df):
+        graph = nx.from_pandas_edgelist(tweet_df, "id", "parent_id")
+        graph.remove_node(0)
+        adj = nx.adjacency_matrix(graph, nodelist=tweet_df["id"].unique()).A
```

Comment:
Note the `.unique()` change in the doc so that dung~ can update in his repo later.